introduce binding_exprt [blocks: #4510]

kroening opened the pull request:
This is a base class that factors out some common functionality of various
expressions that are variable bindings; examples are let, lambda, forall,
exists; future additions might include choose.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening committed:
introduce binding_exprt

This is a base class that factors out some common functionality of various
expressions that are variable bindings; examples are let, lambda, forall,
exists; future additions might include choose.

kroening committed:
quantifiers are variable bindings

This commit makes the quantifier base class use the bindings base class.
Methods and constructor for the special case of quantifying over one
variable are retained.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ce8b26cd60bcedc22a900e899b36675a05caeb46).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/119891946

smowton commented on src/solvers/flattening/boolbv_quantifier.cpp:
```diff
@@ -145,14 +140,14 @@ instantiate_quantifier(const quantifier_exprt &expr, const namespacet &ns)
 
-  const exprt min_i = get_quantifier_var_min(var_expr, re);
-  const exprt max_i = get_quantifier_var_max(var_expr, re);
+  const auto min_i = get_quantifier_var_min(var_expr, re);
```
use `constant_exprt`, nothing else here makes it obvious we're talking about a constant

kroening replied:
changed

smowton reviewed (commented):
Why don't the users (e.g. `typecheck_expr_main`) use the newly-introduced interface instead of the pretty-opaque `expr.op0().op0().op0()`?

kroening committed:
make return type of get_quantifier_var_min obvious

This clarifies that the expression returned by get_quantifier_var_min is a
constant -- request by @smowton.

tautschnig reviewed (changes requested):
These `.op0()` chains are very opaque and make it rather impossible to understand what's going on. And then wouldn't we really also need to make some changes in the back-end now that the structure of those expressions is different? And the commit message talks about lambdas, lets, but only quantifiers received this new treatment. I don't think we should do half-a-job here.

kroening commented:
I've introduce further typing to reduce the `op0()` chains.
Note that we don't have lamba expressions yet, so this PR can't change them.

tautschnig commented:
> Note that we don't have lamba expressions yet, so this PR can't change them.

Well, we have `array_comprehension_exprt` (with `id() == ID_lambda`).

kroening commented:
I'll clean up `array_comprehension_exprt` vs. Church's lamba_exprt in a follow-up PR; this should not be done here.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 190f1c8950762802f3335cd0fadc83a5b4cabcb8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/121976290

tautschnig commented on src/util/mathematical_expr.h:
```diff
@@ -271,6 +271,47 @@ inline function_application_exprt &to_function_application_expr(exprt &expr)
+
+  variablest &variables()
+  {
+    return (variablest &)static_cast<tuple_exprt &>(op0()).operands();
```
Quite likely this isn't the only such place, but this clearly makes `USE_LIST` no longer possible -- but in this case without causing compilation errors. So should we really get rid of `USE_LIST` right now?

kroening replied:
It would be possible to make this work with `USE_LIST`. Is `USE_LIST` sill useful?

tautschnig replied:
The code wouldn't actually compile with `USE_LIST`. If `USE_LIST` were to work, we could change `irept` to have only a single container and some cursor(s) into that. Whether that's more efficient or not I can't tell, because it wouldn't compile...

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 3243b986fb94d6d05a40c540dbef29c99d6c0e2c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/122132439
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 0983a0cd93ff0a2cf533a81afa338cd054fc833a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/122168140
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0983a0cd93ff0a2cf533a81afa338cd054fc833a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/122168140
